servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -226,8 +227,35 @@ private DnsDiscoveryObserver newDiscoveryObserver(final String address) {
     public Publisher<Collection<ServiceDiscovererEvent<InetAddress>>> dnsQuery(final String address) {
         requireNonNull(address);
         return defer(() -> {
-            ARecordPublisher pub = new ARecordPublisher(address, newDiscoveryObserver(address));
-            return inactiveEventsOnError ? recoverWithInactiveEvents(pub, false) : pub;
+            final DnsDiscoveryObserver discoveryObserver = newDiscoveryObserver(address);
+            ARecordPublisher pub = new ARecordPublisher(address, discoveryObserver);
+            Publisher<? extends Collection<ServiceDiscovererEvent<InetAddress>>> events = inactiveEventsOnError ?
+                    recoverWithInactiveEvents(pub, false) :
+                    pub;
+            return discoveryObserver == null ? events : events.beforeFinally(new TerminalSignalConsumer() {
+                    @Override
+                    public void onComplete() {
+                        // this event will never be triggered
+                    }
+
+                    @Override
+                    public void onError(final Throwable cause) {
+                        try {
+                            discoveryObserver.discoveryFailed(cause);
+                        } catch (Throwable err) {
+                            LOGGER.warn("Unexpected exception from observer while reporting error event", err);
```

Comment:
1. Consider adding `cause` as suppressed exception for `err`, similar to `resolutionFailed`:
https://github.com/apple/servicetalk/blob/38d5166d0633da240371a29391bcc29f97b63246/servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java#L784-L790

2. Consider "while reporting error event" -> "while reporting discovery failure" to give users a hint what callback we tried to invoke.